File: packages/react/src/components/AdaptiveMapShell/panel-detents.ts
// The bottom sheet's detents and the rules that move between them: the
// prototype's, driven and measured (docs/pointr-prototype-initial-sheet-
// 2026-09-20.md §1–§2), shared with the iOS and Compose shells. Pure, so
// every rule is tested without a pointer.

/** Where the bottom sheet can rest. */
export type PanelDetent =
  /** A peek: a fifth of the shell, or the content's peek anchor (`panelPeekAnchor`). */
  | "collapsed"
  /** The resting height, where the map and the sheet share the shell. */
  | "medium"
  /** Nearly the whole shell, for reading a long sheet. */
  | "large"
  /** As tall as the sheet's content, between collapsed and large. */
  | "content"
  /** A share of the shell's height. */
  | { fraction: number }
  /** An exact height in pixels. */
  | { height: number };

export const PANEL_DETENT_RULES = {
  collapsed: 0.2,
  medium: 0.54,
  large: 0.94,
  /** A collapsed sheet still has to fit the handle and a header row. */
  collapsedFloor: 112,
  /** …but never takes more than this of a short shell. */
  collapsedCap: 0.4,
  /** An anchored peek stays within these shares of the shell. */
  peekFloor: 0.24,
  peekCap: 0.72,
  /** The margin under a peek anchor. */
  peekMargin: 16,
  /** Outside this range there is either no map or no sheet worth showing. */
  minFraction: 0.12,
  maxFraction: 0.94,
} as const;

export interface PanelDetentMeasures {
  /** The sheet's content height, for the content detent; 0 until measured. */
  contentHeight?: number;
  /** The bottom edge of the content's peek anchor from the sheet's top; 0 when none. */
  peekBottom?: number;
}

const clamp = (value: number, low: number, high: number) =>
  Math.min(Math.max(value, low), high);

/** The collapsed detent when the content marks a peek anchor. */
export function anchoredCollapsedHeight(
  peekBottom: number,
  shellHeight: number,
): number {
  return clamp(
    peekBottom + PANEL_DETENT_RULES.peekMargin,
    shellHeight * PANEL_DETENT_RULES.peekFloor,
    shellHeight * PANEL_DETENT_RULES.peekCap,
  );
}

/** A detent's height in a shell of `shellHeight` pixels. */
export function panelDetentHeight(
  detent: PanelDetent,
  shellHeight: number,
  measures: PanelDetentMeasures = {},
): number {
  const rules = PANEL_DETENT_RULES;
  const largest = shellHeight * rules.large;
  if (detent === "collapsed") {
    if (measures.peekBottom && measures.peekBottom > 0)
      return anchoredCollapsedHeight(measures.peekBottom, shellHeight);
    return Math.min(
      Math.max(shellHeight * rules.collapsed, rules.collapsedFloor),
      shellHeight * rules.collapsedCap,
    );
  }
  if (detent === "medium") return shellHeight * rules.medium;
  if (detent === "large") return largest;
  if (detent === "content") {
    if (!measures.contentHeight || measures.contentHeight <= 0)
      return shellHeight * rules.medium;
    return clamp(
      measures.contentHeight,
      panelDetentHeight("collapsed", shellHeight, measures),
      largest,
    );
  }
  if ("fraction" in detent)
    return (
      shellHeight * clamp(detent.fraction, rules.minFraction, rules.maxFraction)
    );
  return clamp(detent.height, 0, largest);
}

export function panelDetentKey(detent: PanelDetent): string {
  if (typeof detent === "string") return detent;
  return "fraction" in detent
    ? `fraction:${detent.fraction}`
    : `height:${detent.height}`;
}

export function panelDetentEquals(a: PanelDetent, b: PanelDetent): boolean {
  return panelDetentKey(a) === panelDetentKey(b);
}

/** The offered detents in ascending height, one per distinct height. */
export function orderPanelDetents(
  detents: readonly PanelDetent[],
  shellHeight: number,
  measures: PanelDetentMeasures = {},
): PanelDetent[] {
  const seen = new Set<number>();
  return [...detents]
    .sort(
      (a, b) =>
        panelDetentHeight(a, shellHeight, measures) -
        panelDetentHeight(b, shellHeight, measures),
    )
    .filter((detent) => {
      const height = Math.round(
        panelDetentHeight(detent, shellHeight, measures),
      );
      if (seen.has(height)) return false;
      seen.add(height);
      return true;
    });
}

/** The offered detent closest to a height: where a drag snaps once it ends. */
export function nearestPanelDetent(
  detents: readonly PanelDetent[],
  height: number,
  shellHeight: number,
  measures: PanelDetentMeasures = {},
): PanelDetent | undefined {
  let best: PanelDetent | undefined;
  let bestDistance = Infinity;
  for (const detent of orderPanelDetents(detents, shellHeight, measures)) {
    const distance = Math.abs(
      panelDetentHeight(detent, shellHeight, measures) - height,
    );
    if (distance < bestDistance) {
      best = detent;
      bestDistance = distance;
    }
  }
  return best;
}

/** What a drag that starts on the sheet does, decided once at its first move. */
export type PanelDragKind = "content" | "sheet";

/**
 * The prototype's rule: a sideways move is the content's; at the largest
 * detent an upward drag scrolls the content, and a downward one scrolls it
 * back to its top before the sheet moves; below the largest detent every
 * vertical drag moves the sheet.
 */
export function decidePanelDrag(input: {
  dx: number;
  dy: number;
  atLargestDetent: boolean;
  scrollTop: number;
}): PanelDragKind {
  if (Math.abs(input.dx) > Math.abs(input.dy)) return "content";
  if (input.atLargestDetent && (input.dy < 0 || input.scrollTop > 0))
    return "content";
  return "sheet";
}

/** Movement below this reads as a tap rather than a drag. */
export const PANEL_DRAG_SLOP = 6;

export function panelDetentDescription(detent: PanelDetent): string {
  if (detent === "collapsed") return "Collapsed";
  if (detent === "medium") return "Half height";
  if (detent === "large") return "Expanded";
  if (detent === "content") return "Fitted to content";
  if ("fraction" in detent)
    return `${Math.round(detent.fraction * 100)} percent`;
  return `${Math.round(detent.height)} pixels`;
}

File: packages/android/src/main/java/com/kozmos/components/AdaptiveMapShell/PanelDetents.kt
package com.kozmos.components.adaptivemapshell

import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import kotlin.math.abs
import kotlin.math.roundToInt

// The bottom sheet's detents and the rules that move between them: the
// prototype's, driven and measured (docs/pointr-prototype-initial-sheet-
// 2026-09-20.md §1–§2), shared with the iOS and web shells. Pure, so every
// rule is tested without a finger.

/** Where the bottom sheet can rest. */
sealed class KozmosMapPanelDetent {
    /** A peek: a fifth of the shell, or the content's peek anchor (`Modifier.kozmosPanelPeekAnchor()`). */
    data object Collapsed : KozmosMapPanelDetent()

    /** The resting height, where the map and the sheet share the shell. */
    data object Medium : KozmosMapPanelDetent()

    /** Nearly the whole shell, for reading a long sheet. */
    data object Large : KozmosMapPanelDetent()

    /** As tall as the sheet's content, between collapsed and large; medium until measured. */
    data object Content : KozmosMapPanelDetent()

    /** A share of the shell's height. */
    data class Fraction(val value: Float) : KozmosMapPanelDetent()

    /** An exact height. */
    data class Height(val value: Dp) : KozmosMapPanelDetent()

    val description: String
        get() = when (this) {
            Collapsed -> "Collapsed"
            Medium -> "Half height"
            Large -> "Expanded"
            Content -> "Fitted to content"
            is Fraction -> "${(value * 100).roundToInt()} percent"
            is Height -> "${value.value.roundToInt()} dp"
        }

    companion object {
        const val COLLAPSED_FRACTION = 0.2f
        const val MEDIUM_FRACTION = 0.54f
        const val LARGE_FRACTION = 0.94f

        /** A collapsed sheet still has to fit the handle and a header row. */
        val collapsedFloor = 112.dp

        /** …but never takes more than this of a short shell. */
        const val COLLAPSED_CAP = 0.4f

        /** An anchored peek stays within these shares of the shell. */
        const val PEEK_FLOOR = 0.24f
        const val PEEK_CAP = 0.72f

        /** The margin under a peek anchor. */
        val peekMargin = 16.dp

        /** Outside this range there is either no map or no sheet worth showing. */
        const val MIN_FRACTION = 0.12f
        const val MAX_FRACTION = 0.94f

        /** The collapsed detent when the sheet's content marks a peek anchor. */
        fun anchoredCollapsedHeight(peekBottom: Dp, shellHeight: Dp): Dp =
            (peekBottom + peekMargin).coerceIn(shellHeight * PEEK_FLOOR, shellHeight * PEEK_CAP)
    }
}

/** What the shell has measured about the sheet's content. */
data class KozmosPanelMeasures(
    /** The content's own height, for the content detent; zero until measured. */
    val contentHeight: Dp = 0.dp,
    /** The bottom edge of the content's peek anchor from the sheet's top; zero when none. */
    val peekBottom: Dp = 0.dp
)

/** A detent's height in a shell of [shellHeight]. */
fun KozmosMapPanelDetent.height(shellHeight: Dp, measures: KozmosPanelMeasures = KozmosPanelMeasures()): Dp {
    val largest = shellHeight * KozmosMapPanelDetent.LARGE_FRACTION
    return when (this) {
        KozmosMapPanelDetent.Collapsed ->
            if (measures.peekBottom > 0.dp) {
                KozmosMapPanelDetent.anchoredCollapsedHeight(measures.peekBottom, shellHeight)
            } else {
                // Proportional on a tall phone, but never so short on a
                // landscape shell that the handle and header stop fitting.
                maxOf(shellHeight * KozmosMapPanelDetent.COLLAPSED_FRACTION, KozmosMapPanelDetent.collapsedFloor)
                    .coerceAtMost(shellHeight * KozmosMapPanelDetent.COLLAPSED_CAP)
            }
        KozmosMapPanelDetent.Medium -> shellHeight * KozmosMapPanelDetent.MEDIUM_FRACTION
        KozmosMapPanelDetent.Large -> largest
        KozmosMapPanelDetent.Content ->
            if (measures.contentHeight <= 0.dp) {
                shellHeight * KozmosMapPanelDetent.MEDIUM_FRACTION
            } else {
                measures.contentHeight.coerceIn(KozmosMapPanelDetent.Collapsed.height(shellHeight, measures), largest)
            }
        is KozmosMapPanelDetent.Fraction ->
            shellHeight * value.coerceIn(KozmosMapPanelDetent.MIN_FRACTION, KozmosMapPanelDetent.MAX_FRACTION)
        is KozmosMapPanelDetent.Height -> value.coerceIn(0.dp, largest)
    }
}

/** The offered detents in ascending height, one per distinct height. */
fun orderPanelDetents(
    detents: List<KozmosMapPanelDetent>,
    shellHeight: Dp,
    measures: KozmosPanelMeasures = KozmosPanelMeasures()
): List<KozmosMapPanelDetent> {
    val seen = mutableSetOf<Int>()
    return detents
        .sortedBy { it.height(shellHeight, measures).value }
        .filter { seen.add(it.height(shellHeight, measures).value.roundToInt()) }
}

/** The offered detent closest to a height: where a drag snaps once it ends. */
fun nearestPanelDetent(
    detents: List<KozmosMapPanelDetent>,
    height: Dp,
    shellHeight: Dp,
    measures: KozmosPanelMeasures = KozmosPanelMeasures()
): KozmosMapPanelDetent? =
    orderPanelDetents(detents, shellHeight, measures).minByOrNull { abs((it.height(shellHeight, measures) - height).value) }

/** What a drag that starts on the sheet does, decided once at its first move. */
enum class KozmosPanelDragKind {
    /** Left to the content: a sideways move, or a scroll at the largest detent. */
    Content,

    /** Moves the sheet between its detents. */
    Sheet
}

/**
 * The prototype's rule: a sideways move is the content's; at the largest
 * detent an upward drag scrolls the content, and a downward one scrolls it
 * back to its top before the sheet moves; below the largest detent every
 * vertical drag moves the sheet. The nested-scroll connection applies the
 * same rule to a scrollable child, delta by delta.
 */
fun decidePanelDrag(dx: Float, dy: Float, atLargestDetent: Boolean, scrollOffset: Float): KozmosPanelDragKind =
    when {
        abs(dx) > abs(dy) -> KozmosPanelDragKind.Content
        atLargestDetent && (dy < 0f || scrollOffset > 0f) -> KozmosPanelDragKind.Content
        else -> KozmosPanelDragKind.Sheet
    }
